File: python/45. Jump Game II.py
class Solution(object):
    def jump(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums or len(nums) == 1:
            return 0
        start, end, step, maxdis = 0, nums[0], 1, nums[0]
        while end < len(nums) - 1:
            for i in range(start + 1, end + 1):
                maxdis = max(maxdis, nums[i] + i)
            start = end
            end = maxdis
            step += 1
        return step
        

File: cpp/1-50/45. Jump Game II.cpp
class Solution {
public:
	int jump(vector<int>& nums) {
		if (nums.empty())
			return 0;
		int curmax = nums[0], maxpos = nums[0], jumpcnt = 1;
		int size = nums.size();
		for (int i = 1; i <= min(size - 1, maxpos); i++)
		{
			if (i == size - 1)
				return jumpcnt;
			curmax = max(curmax, nums[i] + i);
			if (i == maxpos) {
				maxpos = curmax;
				jumpcnt++;
			}
		}
		return 0;
	}
};